### PR TITLE
Feature Excluded Volume spezialized for Sc and Bcc Bfm

### PR DESCRIPTION
**Pull request by MartinWenge**
Adding two different features for bcc and sc bfm. Now there is some code doubling, but we have high performance Excluded Volume Checks for both Sc and Bcc BFM. 
Adding MoveAddBccMonomer. Again some code doubling with the Sc counterpart,  MoveAddScMonomer, but now systems for bcc and sc bfm can be created using the add monomer moves beeing safer than adding monomers "by hand".
Test provides for the Features, using all kinds of moves.

### CODE REVIEW

**Comment by MartinWenge** [comment]
I still left the Base move in the implementations for generality. This mean, compiler will not complain, if a "wrong" or just unknown move is applied to the feature, if they inherit from MoveBase. We already talked about this, but couldn't find a concluding solution. It is very useful to "ignore" unknown moves, if you are using different kinds of move. But at the moment I didn't see any other moves than the MoveLocal.. ones. Could this change some day?

File: include/LeMonADE/feature/FeatureExcludedVolumeBcc.h [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+ * @brief Forward declaration. Equivalent to FeatureExcludedVolumeBcc< FeatureLattice <bool> >
+ */
+template<class SpecializedClass = FeatureLattice<> > class FeatureExcludedVolumeBcc;
+
+///////////////////////////////////////////////////////////////////////////////
+//DEFINITION OF THE CLASS TEMPLATE   	                                ///////
+//Implementation of the members below					///////
+///////////////////////////////////////////////////////////////////////////////
+
+template<template<typename> class SpecializedClass, typename ValueType>
```
Comment (hrabbel) [inline include/LeMonADE/feature/FeatureExcludedVolumeBcc.h]: Why exactly do we need two template arguments here? "SpecializedClass" is essentially FeatureLattice, right? So why not name it explicitly? Then, ValueType would would just be the value type of the lattice (rename to LatticeValueType ?), and in the required_features_front we could use LeatureLattice<LatticeValueType>
Or would this not work? 
Same applies of course for FeatureExcludedVolumeSc.

**Comment by hrabbel** [comment]
How about explicitly adding a checkMove for MoveLocalBcc in FeatureExcludedVolumeSc (and vice versa), so that the wrong move is forbidden? Or is this unnecessary overhead?

File: include/LeMonADE/feature/FeatureExcludedVolumeSc.h [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+ * @param ingredients A reference to the IngredientsType - mainly the system.
+ */
+/******************************************************************************/
+template<template<typename> class SpecializedClass, typename ValueType>
+template<class IngredientsType>
+void FeatureExcludedVolumeSc< SpecializedClass<ValueType> >::synchronize(IngredientsType& ingredients)
+{
+	//note: the lattice entries are set to 0 before by the 
+	//synchronize function of FeatureLattice
+	std::cout << "FeatureExcludedVolumeSc_T::synchronizing lattice occupation...\n";
```
Comment (hrabbel) [inline include/LeMonADE/feature/FeatureExcludedVolumeSc.h]: I think the "_T" can be removed

File: include/LeMonADE/feature/FeatureExcludedVolumeSc.h [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+template<template<typename> class SpecializedClass, typename ValueType>
+template<class IngredientsType> 
+void FeatureExcludedVolumeSc< SpecializedClass<ValueType> >::fillLattice(IngredientsType& ingredients)
+{
+	const typename IngredientsType::molecules_type& molecules=ingredients.getMolecules();
+	//copy the lattice occupation from the monomer coordinates
+	for(size_t n=0;n<molecules.size();n++)
+	{
+
+		if( ingredients.getLatticeEntry(molecules[n])!=0 )
```
Comment (hrabbel) [inline include/LeMonADE/feature/FeatureExcludedVolumeSc.h]: Better to test for all 8 lattice sites in the if statement here?

File: include/LeMonADE/feature/FeatureExcludedVolumeSc.h (hunk 2 above)
Comment (MartinWenge) [inline include/LeMonADE/feature/FeatureExcludedVolumeSc.h]: To me, adding the "forbidden" moves explicitly to the features looks like the most accurate and safe way to keep an empty/unknown move in the features to enable user specific moves, and at the same time, exclude the wrong lattice move from the excluded volume features. 

Other ideas?

Comment (hrabbel) [inline include/LeMonADE/feature/FeatureExcludedVolumeSc.h]: I agree.

File: include/LeMonADE/feature/FeatureExcludedVolumeBcc.h (hunk 1 above)
Comment (MartinWenge) [inline include/LeMonADE/feature/FeatureExcludedVolumeBcc.h]: I change the typename _SpecializedClass_ to **LatticeClassType** and _ValueType_ to **LatticeValueType**. This seems to be self-explanatory.

File: include/LeMonADE/feature/FeatureExcludedVolumeSc.h (hunk 3 above)
Comment (MartinWenge) [inline include/LeMonADE/feature/FeatureExcludedVolumeSc.h]: What a mistake! I'll fix it immediately.

**Comment by MartinWenge** [comment]
I added the suggestions and fixed the problems. This should fix #4 now.

**Comment by hrabbel** [comment]
I have tested FeatureExcludedVolumeSc with different systems (single chain, cross linked single chain globule, lipid membrane) and together with some of my features (NNInteraction, Umbrella sampling, external potentials). It seems to be working nicely, both with 2^n and other box sizes. So to my opinion wen could merge this.

**Pull request merged**